session.go

```diff
@@ -507,4 +498,17 @@ func (s *session) EMARTT() time.Duration {
 	return s.emaRTT.GetDuration()
 }
 
+type sessionWriter struct {
+	s *session
+}
+
+func (w sessionWriter) Write(b []byte) (int, error) {
+	select {
+	case <-w.s.closeCh:
+		return 0, errorAlreadyClosed
+	case w.s.out <- b:
+		return len(b), nil
```

Comment:
This is the main change: Stop writing if the session is closed.

Reply:
I like the way you encapsulated this as a `Writer`!